improvement(client)!: enhance Jsonable types

jason-ha opened the pull request:
## Breaking Changes

### Alpha

Require first generic parameter for `Jsonable` and `Serializable` types that was previously `any` and resulted in `any`.
Additionally alter types to use new `JsonableTypeWith<>` as result when `any` or `unknown` is specified such that some constraint is actually produced.

### Internal

devtools-core types were broadly broken as its `Primitive` type was updated to respect primitives supported by `Jsonable`.

## Notes

This is a type only product change. Some tests were altered beyond type changes to respect the new constraints.

Some deprecated and experimental uses of [unqualified] `Jsonable` or `Serializable` have been replaced by `any`.
Other uses have been updated to specify a generic parameter, which may be `unknown` and now use `JsonableTypeWith` typing.

Due to recursive `TypeOnly` filter used in type tests, the definition for `JsonableTypeWith` uses interfaces where there is nested typing, which limits recursion. This may cause complications like imprecise typing in the future and need to be switch to pure type alias recursion.
Which means the type test generation will need to be updated to handle such types.

alexvy86 commented on packages/framework/attributor/src/encoders.ts:
```diff
@@ -29,10 +28,10 @@ export const deltaEncoder: TimestampEncoder = {
 		assert(
 			Array.isArray(encoded),
-			0x4b0 /* Encoded timestamps should be an array of nummbers */,
+			0x4b0 /* Encoded timestamps should be an array of numbers */,
```
Let's fix [here](https://github.com/microsoft/FluidFramework/blob/main/packages/runtime/test-runtime-utils/src/assertionShortCodesMap.ts#L779) too. That file would be regenerated during the assert tagging process when we do the next release to match the fixed comment, but might as well do it now.

alexvy86 commented on packages/tools/devtools/devtools-core/src/data-visualization/VisualTree.ts:
```diff
@@ -33,14 +33,14 @@ export enum VisualNodeKind {
  */
 // eslint-disable-next-line @rushstack/no-new-null
-export type Primitive = bigint | number | boolean | null | string | symbol | undefined;
+export type Primitive = number | boolean | null | string | undefined;
```
@Josmithr you're more familiar with this bit of devtools; any thoughts?

jason-ha replied:
Checked with @Josmithr while developing; those others should not be needed as they aren't (shouldn't be) used in DDSes as not serializable (at least not without special logic).
Please do review.

alexvy86 commented on packages/test/test-end-to-end-tests/src/test/benchmark/PasTable.all.spec.ts:
```diff
@@ -47,10 +40,7 @@ describeNoCompat("PAS Test", () => {
-						const sharedString: SharedString = createString(id, dataStoreRuntime);
-						sharedString.insertText(0, "testValue");
-						this.matrix.setCell(i, j, sharedString);
+						this.matrix.setCell(i, j, "testValue");
```
This feels like it's testing a different thing? DDS-in-DDS vs primitive-in-DDS? Not sure it matters much for functionality, since this is a performance test, but would check with the author if there was a reason to do DDS-in-DDS.

DLehenbauer replied:
To a first approximation, the current Tablero component used by Loop works by inserting a SharedString in every matrix cell.  (It's actually a Scriptor document, but similar).

Is the bug that the benchmark inserted 'sharedString' instead of 'sharedString.handle'?

jason-ha replied:
Yes. `.handle` would suffice.
The code also did some funny things with ids. Per the chat there wasn't a need to store SharedString; so I went with simplification. If storing handle is intended - closer mock up; then easy enough to use `.handle` (and I can fix the id collision).
@DLehenbauer, LMK.

jason-ha replied:
Changed to keep SharedString.

alexvy86 commented on packages/runtime/datastore-definitions/src/jsonable.ts:
```diff
@@ -3,19 +3,50 @@
  * Licensed under the MIT License.
  */
 
+/**
+ * Type constraint for types that are serializable as JSON or have a custom
+ * alternate type.
+ *
+ * @remarks
+ * Use JsonableTypeWith<never> for just JSON serializable types.
+ */
+export type JsonableTypeWith<T> =
```
All the new exports probably want to be `@internal` to go along with what @anthony-murphy is doing as part of cleaning up our public API surface.

jason-ha replied:
api-extractor will not let new exports be @internal if they are used by a non-internal type as this is.

alexvy86 reviewed (commented):
The TS-type magic seems overall ok but I'd wait for someone else to give a thumbs up on that part, I don't feel qualified to do it myself for the more complex changes there.

I'm not familiar with experimental/dds/ot so not sure about that one. But experimental/dds/tree is being used by consumers so I think at least for that one we'd want a `.changeset` file to describe the breaking change in release notes? And since it has breaking changes, should this target next (and a separate PR to main announce this upcoming change)? I get in some cases we're just expanding the valid types for input parameters so not worried about those; but for others I'm not clear if some of the changes would be reflected in any way as breaks in consumer code at compile time.

anthony-murphy commented on packages/runtime/datastore-definitions/src/test/types/validateDatastoreDefinitionsPrevious.generated.ts:
```diff
@@ -245,7 +245,7 @@ use_old_InterfaceDeclaration_IFluidDataStoreRuntimeEvents(
     TypeOnly<old.Jsonable>;
 declare function use_current_TypeAliasDeclaration_Jsonable(
-    use: TypeOnly<current.Jsonable>): void;
+    use: TypeOnly<current.Jsonable<any,any>>): void;
```
breaking changes cannot be made in main

jason-ha replied:
Yes, I am not clear on the process with the changes being made. @tylerbutler, `next` is still alive? One more use for Beta 2 at least?

jason-ha replied:
Looks like plenty of changes still happening in `next`. Now I just wish it wasn't days behind so I could retarget more easily.

jason-ha replied:
Looks like `Jsonable` and `Serializable` have been tagged `internal` (impressive containment). Does that mean this could actually go to `main` directly?

Josmithr commented on experimental/dds/tree2/src/test/domains/json/benchmarks.ts:
```diff
@@ -38,7 +37,7 @@ export function sumMap(cursor: ITreeCursor): number {
 }
 
-export function sumDirect(jsonObj: Jsonable): number {
+export function sumDirect(jsonObj: any): number {
```
Could this be typed as a `Record` instead? If not, `unknown`?

jason-ha replied:
`Record` can't be used because it might be given any type.
`Object.values` doesn't handle `unknown`.
This function actually handles `any` and doesn't require the argument to be "Jsonable". The thing it doesn't handle (aka doesn't produce the result I would have expected) is a being given a scalar `number`. That will return 0. In all other cases it finds those numbers.
I originally tried to use a more sophisticated type, but `any` really fits.

DLehenbauer replied:
I think the implied assumption is that jsonObj is an 'object'.

jason-ha replied:
Name certainly implies that but that isn't how it is called, at least not by itself.

Josmithr commented on packages/dds/sequence/src/sharedSequence.ts:
```diff
@@ -31,7 +31,7 @@ export class SubSequence<T> extends BaseSegment {
 		return segment.type === SubSequence.typeString;
 	}
-	public static fromJSONObject<U>(spec: Serializable) {
+	public static fromJSONObject<U>(spec: any) {
```
`unknown`?

jason-ha replied:
`SubSequence` is deprecrated - keeping type the same => `any`.

Josmithr commented on packages/runtime/datastore-definitions/src/jsonable.ts:
```diff
@@ -3,19 +3,50 @@
+	| ArrayLike<JsonableTypeWith<T>>;
+
+/**
+ * @remarks
```
Nit, since this is internal dev facing: 
```suggestion
 * @privateRemarks
```

Josmithr replied:
We may also want to add a public facing note that this is a temporary hack not intended for public consumption.

jason-ha replied:
Is public facing note `@remarks` or something else?

Josmithr replied:
Yeah, `@remarks` is generally reserved for user-facing notes that are not part of the short semantic summary.

Josmithr commented on packages/runtime/datastore-definitions/src/jsonable.ts:
```diff
@@ -32,26 +63,32 @@
- * Using `Jsonable` (with no type parameters) or `Jsonable<any>` is just a type alias for `any`
- * and should not be used if type safety is desired.
+ * Using `Jsonable<unknown>`, or `Jsonable<any>` is just a type alias for
+ * `JsonableType` and should not be used if precise type safety is desired.
```
Nit
```suggestion
 * {@link JsonableType} and should not be used if precise type safety is desired.
```

jason-ha committed:
improvement(client)!: enhance Jsonable types

Require first generic parameter for `Jsonable` and `Serializable` types that was previously `any` and resulted in `any`.
Additionally alter types to use new `JsonableType[With]` as result when `any` or `unknown` is specified such that some constraint is actually produced.

devtools-core types were broadly broken as its `Primitive` type was updated to respect primitives supported by Jsonable.

This is a type only product change. Some tests were altered beyond type changes to respect the new constraints.

Most deprecated and experimental uses of [unqualified] `Jsonable` or `Serializable` have been replaced by `any`.
Other uses have been updated to specify a generic parameter, which may be `unknown` and now use `JsonableTypeWith` typing.

Due to recursive `TypeOnly` filter used in type tests, the definition for `JsonableTypeWith` uses interfaces where there is nested typing, which limits recursion. This may cause complications like imprecise typing in the future and need to be switch to pure type alias recursion.
Which means the type test generation will need to be updated to handle such types.

jason-ha commented:
> The TS-type magic seems overall ok but I'd wait for someone else to give a thumbs up on that part, I don't feel qualified to do it myself for the more complex changes there.
> 
> I'm not familiar with experimental/dds/ot so not sure about that one. But experimental/dds/tree is being used by consumers so I think at least for that one we'd want a `.changeset` file to describe the breaking change in release notes?

experimental/dds/tree (`@fluid-experimental/tree`) doesn't actually have breaking changes as it was using `Serializable` without type parameter and the overall type resolved to `any`. (We really should consider checking that any of our public types are `any`.)
`@fluid-experimental/sharejs-json1` did get a break. We can easily walk that back and use `any` like for tree.

> ... And since it has breaking changes, should this target next (and a separate PR to main announce this upcoming change)? I get in some cases we're just expanding the valid types for input parameters so not worried about those; but for others I'm not clear if some of the changes would be reflected in any way as breaks in consumer code at compile time.

Since `next` is still in use that is updated to be the target.
This is breaking because:
1. we are constricting types that used to be any
2. we require type parameters where there used to be defaults
3. not likely to impact anyone but changed second default on Jsonable
So yes consumers could get build breaks, but no functionality break if they just cast through any. changeset added.

Josmithr commented on experimental/dds/sequence-deprecated/api-report/sequence-deprecated.api.md:
```diff
@@ -188,7 +188,7 @@ export class SparseMatrixFactory implements IChannelFactory {
 
 // @public @deprecated (undocumented)
-export type SparseMatrixItem = Serializable;
+export type SparseMatrixItem = any;
```
Not for this PR, but question for other reviewers: can we delete this and other deprecated aspects of Matrix? There's a fair amount of old cruft here that's been deprecated for a long time.

DLehenbauer replied:
There is fear that this will break ancient documents created with pre-release versions of Loop.

(I'm not sure this is true, but research will be required.)

Josmithr commented on packages/runtime/datastore-definitions/src/serializable.ts:
````diff
@@ -23,4 +23,4 @@ import { Jsonable } from "./jsonable";
  * ```
  */
-export type Serializable<T = any> = Jsonable<T, IFluidHandle>;
+export type Serializable<T> = Jsonable<T, IFluidHandle>;
````
I'm not sure I have a strong opinion here, but would it be convenient to default T to `unknown` here?

Josmithr reviewed (approved):
I left a couple of nitpicks and left a couple of non-urgent questions, but otherwise the changes look good to me. Thanks for improving this space!

jason-ha commented:
#17328 introduces a `readonly` `Serializable` equivalent. I was thinking it could possibly make sense to distinguish to-be-serialized (readonly because serializer shouldn't be changing content) and unserialized cases (which removes the `undefined` possibility since there is no such result), but those could/would be follow-on changes as we see the need. Unsure if that PR means we should be thinking about this sooner.

CraigMacomber commented:
> #17328 introduces a `readonly` `Serializable` equivalent. I was thinking it could possibly make sense to distinguish to-be-serialized (readonly because serializer shouldn't be changing content) and unserialized cases (which removes the `undefined` possibility since there is no such result), but those could/would be follow-on changes as we see the need. Unsure if that PR means we should be thinking about this sooner.

I think it makes sense to land this PR, then land one based on this that adds a read only variant of the new JsonableWithType, then update my isFluidHandle change to use that type, then land it.

DLehenbauer commented on experimental/dds/sequence-deprecated/src/sparsematrix.ts:
```diff
@@ -90,7 +89,7 @@ export class PaddingSegment extends BaseSegment {
  * Use {@link @fluidframework/matrix#SharedMatrix} instead.
  */
-export type SparseMatrixItem = Serializable;
+export type SparseMatrixItem = any;
```
I understand that before your change, Jsonable/Serializable without a generic argument evaluated to the 'any' type.

I thought your change made Jsonable/Serializable<unknown> produce a useful constraint, and it does seem like you use it that that way below...

...So I'm confused as to why you're not using it here?

jason-ha replied:
Because this is deprecated and simply chose to keep the type the same.

DLehenbauer commented on packages/framework/attributor/src/test/attribution/sharedString.attribution.spec.ts:
```diff
@@ -91,34 +89,43 @@ interface FuzzTestState extends BaseFuzzTestState {
+// Type aliases are used here instead of interfaces to avoid ts error 2345:
+//   Index signature for type 'string' is missing in type 'Operation[]'
+// when sent to writeJson which requires Jsonable<unknown>.
+/* eslint-disable @typescript-eslint/consistent-type-definitions */
```
Do you understand why type vs. interface makes a difference?  (I don't, which makes me nervous.)

jason-ha replied:
I think I know. `interfaces` are assumed to be augmented but "simple" types are not. I think this the best comment in the now closed typescript issue: https://github.com/microsoft/TypeScript/issues/15300#issuecomment-657218214

What I boiled it down to in my head whether right or wrong is that with type aliases more likely we are talking about simple data AND we think we'll be getting just what we think.

DLehenbauer commented on packages/framework/attributor/src/test/attribution/sharedString.attribution.spec.ts:
```diff
@@ -362,11 +369,11 @@ function spyOnOperations(baseGenerator: Generator<Operation, FuzzTestState>): {
 }
 
-function writeJson(filepath: string, content: Jsonable) {
+function writeJson(filepath: string, content: Jsonable<unknown>) {
```
In the past, you would want to use the same pattern as readJson:

function writeJson<T>(..., content: Jsonable<T>)

This would infer (if possible) the concrete type of T and recursively check to make sure it doesn't contain forbidden things.

Is Jsonable<unknown> just as good now?

jason-ha replied:
We can write `writeJson<T extends JsonableType>(..., content: Jsonable<T>)` (which I think is the only way that any checking would happen), but I don't think it provides any advantage now over simple `Jsonable<unkown>` aka `JsonableType` content type.

jason-ha replied:
I dug into this some more. Using `writeJson<T>(..., content: Jsonable<T>)` but not `writeJson<T extends JsonableType>(..., content: Jsonable<T>)` will get past the interface vs. type alias issue relating to `Index signature for type 'string' is missing in type 'Foo'.ts(2345)`.
However, doing the same to `readJson` (removing the generic parameter constraint) lets things through that are not correct like having a function on the type. This also really reasserts the asymmetry of serializable to and from string.

I'm placing a deserialized JSON filter in this file as it has immediate need. That type should be moved to datastore-definitions when we have a plan for distinguishing to-serialize and from-serialization types.

cc: @CraigMacomber

jason-ha commented on packages/dds/map/src/test/mocha/map.fuzz.spec.ts:
```diff
@@ -25,7 +25,7 @@ interface Clear {
 	type: "setKey";
 	key: string;
-	value: Jsonable;
+	value: JsonableType;
```
It could be `Jsonable<unknown>` or `JsonableType`, which are now the same. `Jsonable<>` is meant to be a filter for the Jsonable portions of given type param and there is nothing to filter here; so just `JsonableType` seemed most appropriate.

jason-ha replied:
In places that didn't have a use for `Jsonable` like this file, I would fully switch to `JsonableType`. If `Jsonable<>` was otherwise in use I used `Jsonable<unknown>` for symmetry/consistency.

jason-ha replied:
I am not in favor of the `Jsonable` rename. 1) It isn't really an assert. It produces a new type that might be useful. I think an assert would come from checking that `T extends JsonableType`. 2) Rename is more breaking than this change; good existing uses should be fine.
I am open to alternate names for `JsonableType` and `JsonableTypeWith<>`. tree2's types are very similar. Big difference looks like handling of `undefined`. (Would like the datastore-definitions to have better handling for the deserialized typing which does not support `undefined`.) I would not want to use `JsonComaptible` naming unless we reconcile the types. Too confusing otherwise.
I could see changing `Jsonable<>` to `JsonableSubset<>` or `JsonSubset<>` if we think a bigger breaking change is worth it. Then, assuming `Type` is not preferred in type named, `JsonableType` could just be `Jsonable` (type for a value that can be effectively* serialized directly to Json excluding some caveats (* - doesn't mean it would be deserialized exactly (the `undefined` issue))).

jason-ha replied:
I see also that tree2 doesn't do generate type compat testing. Is there "manual" testing for compat?

DLehenbauer replied:
Tree2 has not yet been released / used in production.  (It's our big 2.0 feature.)

If Jsonable<unknown> and JsonableType are interchangeable, I would think it's easier to explain that we have a single type that can do both.

(It's what we tried to do with the original Jsonable<any>, but unfortunately didn't get the result we wanted...  but left it in as documentation.)

jason-ha replied:
`Jsonable` and `JsonableType` are not interchangeable. `JsonableType` is an expression of what of `any` can be serialized to json. `Jsonable` is a utility to get the json serializable portions of a type. A qualified `Jsonable<Foo>` is a specific type. `Jsonable<any | unknown>` maps to `JsonableType`, but no other case should. We can use `Jsonable<unknown>` broadly, but I think it is nicer to just use the thing that it will be (`JsonableType`) rather than having people need to understand that `Jsonable<unknown>` is a special case.
In this file, we are looking at test code (internal) and `JsonableType` as the most clear type makes sense to me.

jason-ha replied:
FWIW I think the most exposure will really be via `Serializable<>`. I did not create a canonical base type for that one as I wasn't seeing so much use. In general, I think we hope to avoid `Serializable<unknown>`.

jason-ha replied:
Well, I looked for uses and this was really the only case. I have plenty more in my prototype, but those may yet become Serializable. So, I'll go for the path of `JsonableType` being an implementation detail for now.

jason-ha replied:
In my prototype the most common use of `JsonableType` is for a generic constraint.
 has:
```typescript
function readJson<T extends Jsonable<unknown>>(filepath: string): Jsonable<T> {
```
but I would instead do
```typescript
function readJson<T extends JsonableType>(filepath: string): Jsonable<T>
// or
function readJson<T extends JsonableType>(filepath: string): T
```
(I haven't explored if there is a meaningful difference in `Jsonable<T>` versus `T` when `T extends `JsonableType`, but I am imagining there is a use when there are distinct types for serializable versus deserialized cases.)

jason-ha replied:
`JsonableType` is no longer in this PR.
I will resolve this conversation, but feel free to unresolve to discuss naming changes that should happen now.

DLehenbauer commented on experimental/dds/tree/api-report/experimental-tree.api.md:
```diff
@@ -731,7 +730,7 @@ export interface ParentData {
 
 // @public
-export type Payload = Serializable;
+export type Payload = any;
```
I assume this is the same deliberate choice to leave "legacy" SharedTree alone.

jason-ha replied:
Yes.

DLehenbauer commented on experimental/framework/data-objects/src/signaler/signaler.ts:
```diff
@@ -7,15 +7,14 @@ import { EventEmitter } from "events";
 // throttling and batching
 
-export type SignalListener = (clientId: string, local: boolean, payload: Jsonable) => void;
+export type SignalListener = (clientId: string, local: boolean, payload: any) => void;
```
This one I would convert.  It's experimental, so not critical...  but also a low-risk opportunity to exercise the new type.

jason-ha replied:
Lowish risk but plenty of overhead. The lack of typing in the signal space is how I started my journey. There isn't any association of specific signals and their payloads. If these were changed to `JsonableType` then all existing listeners will be dealing with that type and not `any`. For example, in example https://github.com/microsoft/FluidFramework/blob/main/examples/apps/presence-tracker/src/FocusTracker.ts:
```typescript
	private readonly onFocusSignalFn = (clientId: string, payload: IFocusSignalPayload) => {
		const userId: string = payload.userId;
		const hasFocus: boolean = payload.focus;
```
and `.userId` and `.focus` "do not exist" on `JsonableType`.
I think we preserve the ugly truth (stop hiding it) in this change. We are hoping not to officially support signals any time soon. We can try to improve the typing here, but I think that is another change.
I am using this new stuff with the new LCEDDS feature prototype. Not worried about getting coverage here (currently).

jason-ha replied:
Looks like typescript can do enough inferences if we genericize this type. `SignalListener<T> ... payload: Jsonable<T>` will work out. None of our examples need any changes with that; so, decent indicator that it isn't terribly disruptive.

DLehenbauer commented on packages/runtime/datastore-definitions/src/jsonable.ts:
```diff
@@ -3,19 +3,54 @@
+	| string
+	| T
+	| Internal_InterfaceOfJsonableTypesWith<T>
+	| ArrayLike<JsonableTypeWith<T>>;
```
We should clarify when to use this type vs. other exported types.

(The simplest story is to say this is an implementation detail and point people at Jsonable.)

DLehenbauer commented on packages/runtime/datastore-definitions/src/jsonable.ts:
```diff
@@ -3,19 +3,54 @@
+ * The TypeOnly filter is not useful for {@link JsonableTypeWith}; so, if type testing improves, this can be removed.
+ */
+export interface Internal_InterfaceOfJsonableTypesWith<T> {
+	[index: string | number]: JsonableTypeWith<T>;
```
@CraigMacomber - I assume this is why the new 'Jsonable<T>' has trouble with interfaces as opposed to types.

An interface won't pattern match '{ [index: string]: ... }' because interfaces do not include '[index: string]: ...' (per https://github.com/microsoft/TypeScript/issues/15300#issuecomment-657218214, as Jason notes above.)

This makes me wonder if we want the old Jsonable<T> (but disallow T=any), plus the new JsonableType.

But it would be better if we could find a fix for interfaces...  Any thoughts?

jason-ha replied:
This is not why `Jsonable` has trouble with interfaces. If this was type alias, the same issue would apply.
This workaround is for the type test system that wraps types with `TypeOnly` even when that isn't helpful... and in the case of recursive types is problematic.

jason-ha committed:
remove JsonableType (as not yet needed)

jason-ha committed:
improvement: add Signaler payload type safety

jason-ha committed:
fix(client): restore test with serialized types
 to interfaces with local introduction of JsonDeserialized.

jason-ha committed:
improvement(client): use generic pattern with Serializable
for correctness and avoids limitation on using interfaces

restore Serial able generic pattern

jason-ha committed:
fix(client): correct types to remove dangerous cast
fromJson only knows how to handle real Arrays of content and not IArrayish. Change content sent to be typed as proper Array as that is how they are constructed anyway.

jason-ha committed:
test(client): coverage for new Jsonable feature
test that T as any and unknown still provide limitations
+ demonstration of limitations

jason-ha committed:
improvement(client): expand Jsonable to handle recursive arrays
without match pattern yeilding infinite recursion
by explicitly matching arrays.
additionally unrecognized primitive types like `symbol` are now filtered to `never` instead of `{}`.

Note: type test TypeOnly limitation prevents filtering ArrayLike types from Jsonable even though those are likely problematic for serialization

jason-ha committed:
improvement(client): cleanup some Serializ/Jsonable<unknown>
to use proper generic <T> pattern
